packages/editor/src/services/EditorStore.ts

```diff
@@ -185,6 +194,102 @@ export class EditorStore {
     this.lastSavedComponentsVersion = val;
   };
 
+  setActiveDataSource = (dataSource: ComponentSchema | null) => {
+    this.activeDataSource = dataSource;
+  };
+
+  setActiveDataSourceType = (dataSourceType: DataSourceType | null) => {
+    this.activeDataSourceType = dataSourceType;
+  };
+
+  getDataSources = () => {
+    const apis: ComponentSchema[] = [];
+    const states: ComponentSchema[] = [];
+
+    this.components.forEach(component => {
+      if (component.type === 'core/v1/dummy') {
+        component.traits.forEach(trait => {
+          if (trait.type === 'core/v1/fetch') {
+            apis.push(component);
+          }
+
+          if (trait.type === 'core/v1/state') {
+            states.push(component);
+          }
+        });
+      }
+    });
+
+    return { apis, states };
+  };
+
+  createDataSource = (type: DataSourceType) => {
```

Comment:
We should make it a new operation. Otherwise `createDataSource` cannot be reverted in once.